vsphere/resource_vsphere_virtual_machine.go

```diff
@@ -260,10 +265,19 @@ func resourceVSphereVirtualMachineRead(d *schema.ResourceData, meta interface{})
 		}
 		return fmt.Errorf("error searching for with UUID %q: %s", id, err)
 	}
+
 	vprops, err := virtualmachine.Properties(vm)
 	if err != nil {
 		return fmt.Errorf("error fetching VM properties: %s", err)
 	}
+
+	err = d.Set("moid", vm.Reference().Value)
```

Comment:
This section looks like it could be optimized a bit, addressing a few things:

* The error you are getting back from `d.Set()` is actually for the state write, not for fetching the MOID, whose path does not yield an error.
* Since you are setting primitives, errors are not necessarily useful as the risk of there being a state write error on such an attribute is negligible.
* The log message that tells you what the MOID is could possibly be a little more informative.

If you can update it so that it resembles what's below that'd be great:

```go
moid := vm.Reference().Value
log.Printf("[DEBUG] MOID for VM %q is %q", vm.InventoryPath, moid)
d.Set("moid", moid)
```

Reply:
thanks @vancluever if its like u are telling me, u are right and it makes no sense. :-) So I update it! I will test it in the next half hour.